File: api/views/playlists.py
from api.serializers.playlists import PlaylistSerializer
from api.views import create_response
from django.shortcuts import get_object_or_404
from rest_framework.decorators import api_view

from playlists.models import UserPlaylist
from artists.models import Song
from django.contrib.auth import get_user_model

USER_MODEL = get_user_model()

@api_view(['get'])
def get_user_playlists(request, **kwargs):
    queryset = UserPlaylist.objects.filter(author__username='zadigo')
    serializers = PlaylistSerializer(instance=queryset, many=True)
    return create_response(serializer=serializers)


@api_view(['post'])
def add_song_to_playlist(request, pk, **kwargs):
    playlist = get_object_or_404(UserPlaylist, id=pk)
    song = get_object_or_404(Song, id=request.data.get('song_id'))
    playlist.songs.add(song)
    return create_response(data={'state': 'True'})


@api_view(['post'])
def create_playlist(request, **kwargs):
    # TODO: Authentify the user that sends
    # the request by using a token
    user = get_object_or_404(USER_MODEL, id=1)
    
    last_playlist = UserPlaylist.objects.last()
    
    name = f'Playlist n°{last_playlist.id + 1}'
    playlist = UserPlaylist.objects.create(author=user, name=name)
    serialized_playlist = PlaylistSerializer(instance=playlist)
    
    return create_response(serializer=serialized_playlist)


@api_view(['post', 'delete'])
def delete_playlist(request, pk, **kwargs):
    # TODO: Authentify the user on the backend
    user = None
    # playlist = get_object_or_404(UserPlaylist, id=request.data.get('playlist_id'), user=user)
    playlist = get_object_or_404(UserPlaylist, id=pk)
    playlist.delete()
    return create_response(data={ 'id': playlist.id })

File: front/src/plugins/api/playlists.js
var getAll = (_axios) => {
    return (userId) => {
        return _axios({
            method: 'get',
            url: '/playlists',
            data: { user_id: userId }
        })
    }
}

var addSong = (_axios) => {
    return (playlistId, songId) => {
        return _axios({
            method: 'post',
            url: `/playlists/${playlistId}/add`,
            data: { song_id: songId }
        })
    }
}

var removeSong = (_axios) => {
    return (playlistId, songId) => {
        return _axios({
            method: 'delete',
            url: `/playlists/${playlistId}/remove`,
            data: { song_id: songId }
        })
    }
}

var deletePlaylist = (_axios) => {
    return (playlistId) => {
        return _axios({
            method: 'delete',
            url: `/playlists/${playlistId}/delete`
        })
    }
}

var changeName = (_axios) => {
    return (playlistId, newName) => {
        return _axios({
            method: 'post',
            url: `/playlists/${playlistId}`,
            data: { name: newName }
        })
    }
}

var changeSorting = (_axios) => {
    return (playlistId, method) => {
        return _axios({
            method: 'post',
            url: `/playlists/${playlistId}`,
            data: { method: method }
        })
    }
}

var createPlaylist = (_axios) => {
    return () => {
        return _axios({
            method: 'post',
            url: `/playlists/create`
        })
    }
}

export default (_axios) => ({
    add: addSong(_axios),
    remove: removeSong(_axios),
    
    all: getAll(_axios),
    create: createPlaylist(_axios),
    delete: deletePlaylist(_axios),
    
    changeSort: changeSorting(_axios),
    changeName: changeName(_axios)
})
